packages/std/src/imports.rs

```diff
@@ -69,21 +69,31 @@ impl ReadonlyStorage for ExternalStorage {
         end: Option<&[u8]>,
         order: Order,
     ) -> Box<dyn Iterator<Item = KV>> {
-        // start and end (Regions) must remain in scope as long as the start_ptr / end_ptr do
-        // thus they are not inside a block
-        let start = start.map(|s| build_region(s));
-        let start_ptr = match start {
-            Some(reg) => &*reg as *const Region as u32,
```

Comment:
Just out of curiosity, I wonder why these fail in the vm context.

Reply:
They return something, but likely a pointer to stack, which is overwritten, not a pointer to heap.

Reply:
I see. The `db_scan` is out of scope!

Reply:
what do you mean by out of scope?
Looking at this diff it's not clear to me what the difference is between the two versions.

Reply:
Me either. But the compiler does something so they work very different.

Check out the `WIP: lots of debug println` commit and run it, and then switch out what is commented out. This is odd behavior, but we are grabbing a result that was created in a closure and dereferenced inside an Option, and I am assuming something happens a bit odd, so we get a stack pointer not a heap pointer.

Reply:
Honestly, `&*reg as *const Region as u32` is basically a magical incantation to me. It somehow gets a pointer to the content of an option (which is itself boxed).

Reply:
> This is odd behavior, but we are grabbing a result that was created in a closure and dereferenced inside an Option, and I am assuming something happens a bit odd, so we get a stack pointer not a heap pointer.

That was what I meant by "out of scope". I wonder if the `unsafe` blocks play a part too. In the sense that if they weren't there, the compiler would throw an error or warning about the validity / lifetime of that pointer dereference.

It's an interesting thing to experiment with and understand clearly, that's for sure.

Reply:
As I now understand it: `&*reg` is a reference to a value on the stack (the `*` is unboxing the value), which is then converted to a pointer (`as *const Region`) which is then converted to an unsigned integer.

And `unsafe` (the keyword) doesn't have much to do with it. After casting or converting to a raw pointer, you're basically on your own. Not to talk about casting to an integer.

Reply:
I guess there are no safe (in the sense of compiler warning and such) ways to do these mapping between languages / frameworks.

Reply:
OK, I was looking into this, and the alternative way to do it, if you cannot avoid using a raw pointer, is to use `Box::into_raw`, which avoids unboxing. The advantage is that the returned pointer is to the original value on the heap. So, it avoids copying of the value to the stack during unboxing, and the scoping issues.

The code also ends being somewhat clearer:

```
        let start_ptr = match start {
            Some(reg) => Box::into_raw(reg) as u32,
            None => 0,
        };
        let end = end.map(|e| build_region(e));
        let end_ptr = match end {
            Some(reg) => Box::into_raw(reg) as u32,
            None => 0,
        };
```

Now, after `into_raw` you're dealing with a raw (therefore unmanaged) pointer. Being on the heap, it's your responsibility to release it. So, it has to be manually released with something like:
```
        if start_ptr != 0 {
            unsafe { drop(Box::from_raw(start_ptr as *mut Region)); }
        }
        if end_ptr != 0 {
            unsafe { drop(Box::from_raw(end_ptr as *mut Region)); }
        }
```

This can still be improved a bit, by avoiding the cast tu `u32`in the match (that is, casting in the `db_scan` parameters), in order to avoid recasting to `*Region`.

Reply:
Something like this:
```
        let start = start.map(|s| build_region(s));
        let start_ptr = match start {
            Some(reg) => Box::into_raw(reg),
            None => null_mut(),
        };
        let end = end.map(|e| build_region(e));
        let end_ptr = match end {
            Some(reg) => Box::into_raw(reg),
            None => null_mut(),
        };
        let order = order as i32;

        let iterator_id = unsafe { db_scan(start_ptr as u32, end_ptr as u32, order) };
        let iter = ExternalIterator { iterator_id };
        if start_ptr != null_mut() {
            unsafe { drop(Box::from_raw(start_ptr)); }
        }
        if end_ptr != null_mut() {
            unsafe { drop(Box::from_raw(end_ptr)); }
        }
        Box::new(iter)
```

Reply:
Nice suggestion. Do you want to make a branch with those changes and see if it passes tests?

I actually kind of like the current way as it is more explicit and fewer if statments. But the Box::into_raw is a nice addition as well.

You can make a PR proposing this as a cleaner logic and we can iterate on it and see if anyone has better ideas. I really wanted to get the thing fixed first of all, so iterators work in the contracts, but happy for a refined version in the future. Whatever is easiest to read really.

Reply:
Sure. I already tested it, and it works. I'll make a PR.